hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPServiceActor.java

```diff
@@ -1304,12 +1305,18 @@ long scheduleBlockReport(long delay, boolean isRegistration) {
      * Numerical overflow is possible here.
      */
     void scheduleNextBlockReport() {
+      scheduleNextBlockReport(false);
+    }
+
+    void scheduleNextBlockReport(boolean forceFullBr) {
       // If we have sent the first set of block reports, then wait a random
       // time before we start the periodic block reports.
       if (resetBlockReportTime) {
         nextBlockReportTime.getAndSet(monotonicNow() +
             ThreadLocalRandom.current().nextInt((int) (blockReportIntervalMs)));
         resetBlockReportTime = false;
+      } else if (forceFullBr) {
+        nextBlockReportTime.getAndSet(monotonicNow() + blockReportIntervalMs);
```

Comment:
If many datanodes of a large cluster is triggered in batches, the FBR time of these datanodes will be concentrated in the future, which may cause great pressure on NN. Maybe we also need to add random values.

Reply:
I think that in most of cases, we want to know the exactly nextBlockReportTime if we trigger block report by force.
As for trigger full block report in batches, I think it is a dangerous behavior because it may cause some datanode heartbeat timeout.

Reply:
> I think that in most of cases, we want to know the exactly nextBlockReportTime if we trigger block report by force. As for trigger full block report in batches, I think it is a dangerous behavior because it may cause some datanode heartbeat timeout.

I mean, if we trigger the FBR on some nodes in a short period of time, it will affect the original randomness.

Reply:
@liubingxing @tomscut 
Thanks for the discussion. IMHO, I prefer to use a random time to keep the randomness.
I think HDFS-15167 causes this problem. Before HDFS-15167, `resetBlockReportTime` is true when triggering full block report by force, and it used a random time.